Command rework

keraldi opened the pull request:
Commands rewrite.

- Moved Teleports.java to a more appropiate location (rituals/portal -> teleport)
- Added a ritual command to create/repair rituals.
- Added a teleposer command to list, teleport to teleposers or their focus locations and to remove valid teleposers (teleposers that contain a bound focus).
- Fixed some missing strings (for the english version)

Commit:
Network part finished.

Commit:
Should be more reasonable now

Commit:
This should be good enough.

Commit:
Orb finished, needs strings

Commit:
Bind finished. Needs strings.

Commit:
Reformat & Help subcommand

Commit:
Cleanup, strings, no negative amounts

Commit:
Removed TODOs

Commit:
Added missing MaxTier check for Blood Orbs.
Added TODO: Test with custom Blood Orbs.

Commit:
Ritual commands finished.
Check for valid placement might be optimized. (TODO)

Commit:
Access modifiers, moved TODO

Commit:
Added TODOs for localized strings

Commit:
DrainUtils postponed until the necessary functionality is available with SoulTickets (telling SoulTicket network from soul ticket, a list of all registered soul tickets per network)

Iorce commented:
If this gets merged, I'll make a suggestion for a ritual repair/remove item.

Commit:
Replaced all occurrences of TextHelper with TextComponentTranslation in the commands section

Commit:
- Moved Teleports.java to teleport package
- added teleposer command
- added missing strings
- cleanup

keraldi commented:
Update. Ready for review.

Commit:
Fixed spelling of "Successful(ly)"

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandBind.java:
```diff
@@ -3,25 +3,35 @@
+        addSubcommand(new CommandTreeHelp(this));
+    }
+
+    public Object getInfo() {
```
Why `Object`. Why not `String`?

keraldi replied:
getInfo() for commands.bloodmagic.network.get is Integer.

I made it Object to have more leeway with help strings using multiple components.

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandBind.java:
```diff
@@ -3,25 +3,35 @@
+import net.minecraftforge.server.command.CommandTreeHelp;
+
+public class SubCommandBind extends CommandTreeBase {
+    public String help = new TextComponentTranslation("commands.bloodmagic.bind.help").getFormattedText();
```
This will never be re-translated after the language is changed in-game.

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandBind.java:
```diff
@@ -3,25 +3,35 @@
     @Override
     public String getUsage(ICommandSender commandSender) {
-        return TextHelper.localizeEffect("commands.bloodmagic.bind.usage");
+        return new TextComponentTranslation("commands.bloodmagic.bind.usage").getFormattedText() + "\n" + help;
```
`getUsage(...)` expects a localization key.

keraldi replied:
It was the only way to have a halfway decent display of the help string <_<

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandNetwork.java:
```diff
@@ -2,182 +2,215 @@
     @Override
     public String getUsage(ICommandSender commandSender) {
-        return TextHelper.localizeEffect("commands.bloodmagic.network.usage");
+        return new TextComponentTranslation("commands.bloodmagic.network.usage").getFormattedText();
```
`getUsage(...)` expects a localization key.

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandNetwork.java:
```diff
@@ -2,182 +2,215 @@
-                EntityPlayer player = CommandBase.getPlayer(server, commandSender, args[1]);
+        @Override
+        public String getUsage(ICommandSender sender) {
+            return new TextComponentTranslation("commands.bloodmagic.network." + getName() + ".usage").getFormattedText() + "\n" + help;
```
`getUsage(...)` expects a localization key.

Iorce replied:
It returns localization keys and they're getting translated.

TehNut replied:
You are returning a localized value. It expects a ***key***.

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandNetwork.java:
```diff
@@ -2,182 +2,215 @@
+
+                for (SoulTicket i : network.getTicketHistory())
+                    sender.sendMessage(i.getDescription());
+            sender.sendMessage(new TextComponentTranslation("commands.bloodmagic.network.ticket.success", player.getDisplayName().getFormattedText()));
```
Let the component handle the player name. Just provide `player.getDisplayName()`.

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandOrb.java:
```diff
@@ -1,107 +1,116 @@
     @Override
     public String getUsage(ICommandSender commandSender) {
-        return TextHelper.localizeEffect("commands.bloodmagic.orb.usage");
+        return new TextComponentTranslation("commands.bloodmagic.orb.usage").getFormattedText();
```
`getUsage(...)` expects a localization key.

TehNut replied:
This is still an issue.

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandOrb.java:
```diff
@@ -1,107 +1,116 @@
-    public void execute(MinecraftServer server, ICommandSender commandSender, String[] args) throws CommandException {
-        if (args.length > 0) {
+    abstract class OrbCommand extends CommandTreeBase {
+        public String help = new TextComponentTranslation("commands.bloodmagic.orb." + getName() + ".help", getInfo()).getFormattedText();
```
Again, you cannot pre-translate like this. Localization must be done on the fly.

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandOrb.java:
```diff
@@ -1,107 +1,116 @@
-            }
+        @Override
+        public String getUsage(ICommandSender sender) {
+            return new TextComponentTranslation("commands.bloodmagic.orb." + getName() + ".usage").getFormattedText() + "\n" + help;
```
`getUsage(...)` expects a localization key.

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandRitual.java:
```diff
@@ -0,0 +1,158 @@
+    class RitualCreate extends CommandTreeBase {
+        public RitualCreate() {
+            for (Ritual i : BloodMagic.RITUAL_MANAGER.getRituals()) {
+                addSubcommand(new CommandTreeBase() {
```
haha no

There will not be a command for every single ritual.

keraldi replied:
The reason I did this is because it allows command completion via tab. If there's a downside to that, I'd like to know :)

It also automatically includes modded rituals (for the completion).

TehNut replied:
Did `getTabCompletions` stop existing or something?

keraldi replied:
Nope, I didn't know it exists, I'll look into it.

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandTeleposer.java:
```diff
@@ -0,0 +1,326 @@
+import java.util.Set;
+
+public class SubCommandTeleposer extends CommandTreeBase {
+    public static Set<TileTeleposer> teleposerSet = new HashSet<>(); //contains "valid" teleposers (teleposers with focus), teleposers check themselves every 100 ticks.
```
If this is a constant, it should be `final`.

TehNut commented on src/main/java/WayofTime/bloodmagic/tile/TileTeleposer.java:
```diff
@@ -49,13 +51,26 @@ public NBTTagCompound serialize(NBTTagCompound tagCompound) {
+                ItemStack focusStack = getStackInSlot(0);
+                if (!focusStack.isEmpty()) {
+                    ItemTelepositionFocus focus = (ItemTelepositionFocus) focusStack.getItem();
+                    Binding binding = focus.getBinding(focusStack);
```
```java
if (focus.getBinding(focusStack) != null)
    SubCommandTeleposer.teleposerSet.add(this);
```

TehNut commented on src/main/java/WayofTime/bloodmagic/util/helper/RitualHelper.java:
```diff
@@ -117,4 +119,135 @@ else if (tile != null && tile.hasCapability(RUNE_CAPABILITY, null)) {
         }
     }
+
+    public static void removeValidRituals(TileMasterRitualStone tile) {
```
What even is this method? The MRS knows what ritual it has active, use that.

keraldi replied:
The MRS knows what ritual it has but this can remove remains from a non-activated ritual (i.e. something is in the way when using the Ritual Diviner) and it can remove rituals that are complete but have not been activated.

TehNut commented on src/main/java/WayofTime/bloodmagic/util/helper/RitualHelper.java:
```diff
@@ -117,4 +119,135 @@ else if (tile != null && tile.hasCapability(RUNE_CAPABILITY, null)) {
+    }
+
+
+    public static Pair<Ritual, EnumFacing> getRitualFromRuins(TileMasterRitualStone tile) {
```
Again... the MRS already knows this information.

keraldi replied:
This is also supposed to be able to be used on "Rituals" that have never been activated.

TehNut reviewed (changes requested):
First pass

Commit:
getUsage() now returns translation keys.
getInfo() is now an explicit String

ritual creation command now has proper tab completions

help is an additional argument with "-h" or "?"

cleanup

Commit:
teleposerSet final
cleanup.

keraldi commented:
Done.

keraldi commented:
#1428

Commit:
Removed ritual removal command

Signed-off-by: tobias <angryaeon@icloud.com>

Commit:
Check if the tile has a ritual first

Signed-off-by: tobias <angryaeon@icloud.com>

keraldi commented:
update

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandBind.java:
```diff
@@ -3,25 +3,39 @@
+    }
+
+    public String getHelp() {
+        return new TextComponentTranslation("commands.bloodmagic.bind.help").getFormattedText();
```
Pretty sure this is also supposed to just be the localization key.

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandBind.java:
```diff
@@ -30,49 +44,49 @@ public int getRequiredPermissionLevel() {
-    public void execute(MinecraftServer server, ICommandSender commandSender, String[] args) throws CommandException {
-        if (commandSender.getEntityWorld().isRemote)
+    public void execute(MinecraftServer server, ICommandSender sender, String[] args) throws CommandException {
+        if (args.length == 1 && (args[0].equals("?") || args[0].equals("-h"))) {
```
`-h` -> `help`

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandBind.java:
```diff
@@ -30,49 +44,49 @@ public int getRequiredPermissionLevel() {
+                    else
+                        player = getPlayer(server, sender, args[0]);
+            if (bind) {
+                if (binding != null && binding.getOwnerName().equals(player.getName())) {
```
`binding != null` is redundant. If it were null, `bind` would be `false`.

Commit:
A bit more optimisation

Signed-off-by: tobias <angryaeon@icloud.com>

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandNetwork.java:
```diff
@@ -2,182 +2,246 @@
+        }
+
+        @Override
+        public void execute(MinecraftServer server, ICommandSender sender, String[] args) throws CommandException {
```
Instead of setting these here and calling `super` in each sub command, make this method `final` and create a new abstract execution method that this one calls.

This method can also handle the help printing.

keraldi replied:
I am not quite sure what you mean, could you please elaborate?

TehNut replied:
```
public final execute (...) 
    // help handler
    // set stuff
    subExecute(...)

protected abstract subExecute (...)
```

When you extend the generic sub command class, you override the abstract method and only perform the unique behavior. The final method will handle the common stuff

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandNetwork.java:
```diff
@@ -2,182 +2,246 @@
+
+        @Override
+        public void execute(MinecraftServer server, ICommandSender sender, String[] args) throws CommandException {
+            if (args.length == 1 && (args[0].equals("?") || args[0].equals("-h"))) {
```
Move this help bit to super.

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandNetwork.java:
```diff
@@ -2,182 +2,246 @@
+
+        @Override
+        public void execute(MinecraftServer server, ICommandSender sender, String[] args) throws CommandException {
+            if (args.length == 1 && (args[0].equals("?") || args[0].equals("-h"))) {
```
`-h` -> `help`

Also, move this help bit to super.

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandNetwork.java:
```diff
@@ -2,182 +2,246 @@
+        @Override
+        public void execute(MinecraftServer server, ICommandSender sender, String[] args) throws CommandException {
+            if (args.length == 1 && (args[0].equals("?") || args[0].equals("-h"))) {
+                sender.sendMessage(new TextComponentString(getHelp()));
```
Move this help bit to super.

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandNetwork.java:
```diff
@@ -2,182 +2,246 @@
+
+        @Override
+        public void execute(MinecraftServer server, ICommandSender sender, String[] args) throws CommandException {
+            if (args.length == 1 && (args[0].equals("?") || args[0].equals("-h"))) {
```
Move this help bit to super.

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandNetwork.java:
```diff
@@ -2,182 +2,246 @@
-                }
+        @Override
+        public void execute(MinecraftServer server, ICommandSender sender, String[] args) throws CommandException {
+            if (args.length == 1 && (args[0].equals("?") || args[0].equals("-h"))) {
```
Move this help bit to super.

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandNetwork.java:
```diff
@@ -2,182 +2,246 @@
-                }
+        @Override
+        public void execute(MinecraftServer server, ICommandSender sender, String[] args) throws CommandException {
+            if (args.length == 1 && (args[0].equals("?") || args[0].equals("-h"))) {
```
Move this help bit to super.

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandNetwork.java:
```diff
@@ -2,182 +2,246 @@
-        public abstract void run(EntityPlayer player, ICommandSender sender, boolean displayHelp, String... args);
+        @Override
+        public void execute(MinecraftServer server, ICommandSender sender, String[] args) throws CommandException {
+            if (args.length == 1 && (args[0].equals("?") || args[0].equals("-h"))) {
```
Move this help bit to super.

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandOrb.java:
```diff
@@ -1,107 +1,128 @@
+        }
+
+        public String getHelp() {
+            return new TextComponentTranslation("commands.bloodmagic.orb." + getName() + ".help", getInfo()).getFormattedText();
```
This expects a localization key

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandOrb.java:
```diff
@@ -1,107 +1,128 @@
+
+        @Override
+        public void execute(MinecraftServer server, ICommandSender sender, String[] args) throws CommandException {
+            if (args.length == 1 && (args[0].equals("?") || args[0].equals("-h"))) {
```
`-h` -> `help`

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandOrb.java:
```diff
@@ -1,107 +1,128 @@
+        }
+
+        @Override
+        public void execute(MinecraftServer server, ICommandSender sender, String[] args) throws CommandException {
```
Same here as for network. Same for all the sub commands help handling being moved to here.

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandRitual.java:
```diff
@@ -0,0 +1,144 @@
+            if (args.length == 0) {
+                sender.sendMessage(new TextComponentTranslation("commands.blooodmagic.ritual.create.noRitual"));
+                return;
+            } else if (args.length == 2 && (args[1].equals("-h") || args[1].equals("?"))) {
```
`-h` -> `help`

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandTeleposer.java:
```diff
@@ -0,0 +1,353 @@
+        @Override
+        public void execute(MinecraftServer server, ICommandSender sender, String[] args) throws CommandException {
+            if (args.length == 1)
+                if (args[0].equals("?") || args[0].equals("-h")) {
```
I think I've said this enough

TehNut commented on src/main/java/WayofTime/bloodmagic/command/sub/SubCommandTeleposer.java:
```diff
@@ -0,0 +1,353 @@
+        }
+
+        @Override
+        public void execute(MinecraftServer server, ICommandSender sender, String[] args) throws CommandException {
```
Same as the other 2 classes designed like this.

TehNut commented on src/main/java/WayofTime/bloodmagic/tile/TileTeleposer.java:
```diff
@@ -49,13 +51,25 @@ public NBTTagCompound serialize(NBTTagCompound tagCompound) {
 
             previousInput = currentInput;
+
+            if (ticksExisted % 100 == 0) {
```
`world.getTotalWorldTime()`

Commit:
Cleanup part 1

Signed-off-by: tobias <angryaeon@icloud.com>

Commit:
Cleanup part 2

Signed-off-by: tobias <angryaeon@icloud.com>

Commit:
Part 3

Signed-off-by: tobias <angryaeon@icloud.com>

Commit:
Part 4

Signed-off-by: tobias <angryaeon@icloud.com>

Commit:
Updated language file to reflect cleanup & continuity changes.

Signed-off-by: tobias <angryaeon@icloud.com>

Commit:
Change to use an abstract class that gets called instead of calling super on overriden execute() for commands

Signed-off-by: tobias <angryaeon@icloud.com>

keraldi commented:
Done.

Commit:
Use player facing for ritual creation.

Signed-off-by: tobias <angryaeon@icloud.com>
